src/material-experimental/mdc-form-field/form-field.ts

```diff
@@ -0,0 +1,485 @@
+  @ContentChild(MatFormFieldControl) _formFieldControl: MatFormFieldControl<any>;
+  @ContentChildren(MatPrefix, {descendants: true}) _prefixChildren: QueryList<MatPrefix>;
+  @ContentChildren(MatSuffix, {descendants: true}) _suffixChildren: QueryList<MatSuffix>;
+  @ContentChildren(MatError, {descendants: true}) _errorChildren: QueryList<MatError>;
+  @ContentChildren(MatHint, {descendants: true}) _hintChildren: QueryList<MatHint>;
+
+  /** Whether the required marker should be hidden. */
+  @Input() hideRequiredMarker: boolean = false;
+
+  /** The color palette for the form-field. */
+  @Input() color: ThemePalette = 'primary';
+
+  /** Whether the label should always float or float as the user types. */
+  @Input()
+  get floatLabel(): FloatLabelType {
+    return this._floatLabel || (this._labelOptions && this._labelOptions.float)
+        || DEFAULT_FLOAT_LABEL;
+  }
+  set floatLabel(value: FloatLabelType) {
+    if (value !== this._floatLabel) {
+      this._floatLabel = value;
+      // For backwards compatibility. Custom form-field controls or directives might set
+      // the "floatLabel" input and expect the form-field view to be updated automatically.
+      // e.g. autocomplete trigger. Ideally we'd get rid of this and the consumers would just
+      // emit the "stateChanges" observable. TODO(devversion): consider removing.
+      this._changeDetectorRef.markForCheck();
+    }
+  }
+  private _floatLabel: FloatLabelType;
+
+  /** The form-field appearance style. */
+  @Input()
+  get appearance(): MatFormFieldAppearance { return this._appearance; }
+  set appearance(value: MatFormFieldAppearance) {
+    this._appearance = value || (this._defaults && this._defaults.appearance) || DEFAULT_APPEARANCE;
+  }
+  private _appearance: MatFormFieldAppearance = DEFAULT_APPEARANCE;
+
+  /** Text for the form field hint. */
+  @Input()
+  get hintLabel(): string { return this._hintLabel; }
+  set hintLabel(value: string) {
+    this._hintLabel = value;
+    this._processHints();
+  }
+  private _hintLabel = '';
+
+  // Unique id for the hint label.
+  _hintLabelId: string = `mat-hint-${nextUniqueId++}`;
+
+  // Unique id for the internal form field label.
+  _labelId = `mat-form-field-label-${nextUniqueId++}`;
+
+  /** Whether the Angular animations are enabled. */
+  _animationsEnabled: boolean;
+
+  /** State of the mat-hint and mat-error animations. */
+  _subscriptAnimationState: string = '';
+
+  /** Gets the current form field control */
+  get _control(): MatFormFieldControl<any> {
+    return this._explicitFormFieldControl || this._formFieldControl;
+  }
+  set _control(value) { this._explicitFormFieldControl = value; }
+
+  private _destroyed = new Subject<void>();
+  private _isFocused: boolean|null = null;
+  private _explicitFormFieldControl: MatFormFieldControl<any>;
+  private _foundation: MDCTextFieldFoundation;
+  private _adapter: MDCTextFieldAdapter = {
```

Comment:
I tried that for the slider, but IIRC it became really cumbersome due to the private members. For the form-field it seems to work out nicely and I don't need to change any member visibility.

Good idea!

Reply:
Reporting back. This doesn't work because MDC requires the adapter methods to be enumerable. Otherwise they will be overwritten with the default methods since they use `Object.assing`.

```
/** Adapter for the MDC text-field used by MDC-based mat-form-field. */
class MatFormFieldMdcAdapter implements MDCTextFieldAdapter {
  constructor(private _formField: MatFormField) {}

  /** Gets the MDC text-field element. */
  private _getTextFieldElement(): HTMLElement {
    return this._formField._textField.nativeElement;
  }

  /** Adds a class to the text-field element. */
  addClass(className: string) {
    this._getTextFieldElement().classList.add(className);
  }

  /** Removes a class from the text-field element. */
  removeClass(className: string) {
    this._getTextFieldElement().classList.remove(className);
  }

  /** Gets whether a given class is set on the text-field element. */
  hasClass(className: string) {
    return this._getTextFieldElement().classList.contains(className);
  }

  /** Gets whether the text-field has a floating label. */
  hasLabel() {
    return this._formField._hasFloatingLabel();
  }

  /** Gets whether the text-field is currently focused */
  isFocused() {
    return this._formField._control.focused;
  }

  /** Gets whether the text-field has a notched-outline. */
  hasOutline() {
    return this._formField._hasOutline();
  }

  /**
   * Gets the width of the floating label. Width measurements are
   * performed from within the MDCFloatingLabel component.
   */
  getLabelWidth() {
    return this._formField._floatingLabel ? this._formField._floatingLabel.getWidth() : 0;
  }

  /** Opens the notched-outline with the specified label width. */
  notchOutline(labelWidth: number) {
    if (this._formField._notchedOutline) {
      this._formField._notchedOutline.notch(labelWidth);
    }
  }

  /** Closes the notched-outline. */
  closeOutline() {
    if (this._formField._notchedOutline) {
      this._formField._notchedOutline.closeNotch();
    }
  }

  /** Activates the line-ripple. Usually called by the foundation on focus. */
  activateLineRipple() {
    if (this._formField._lineRipple) {
      this._formField._lineRipple.activate();
    }
  }

  /** Deactivates the line-ripple. Usually called by the foundation on blur. */
  deactivateLineRipple() {
    if (this._formField._lineRipple) {
      this._formField._lineRipple.deactivate();
    }
  }

  // MDC text-field will call this method on focus, blur and value change. It expects us
  // to update the floating label state accordingly. Though we make this a noop because we
  // want to react to floating label state changes through change detection. Relying on this
  // adapter method would mean that the label would not update if the custom form-field control
  // sets "shouldLabelFloat" to true, or if the "floatLabel" input binding changes to "always".
  floatLabel() {}

  // Label shaking is not supported yet. It will require a new API for form field
  // controls to trigger the shaking. This can be a feature in the future.
  // TODO(devversion): explore options on how to integrate label shaking.
  shakeLabel() {}

  // The foundation tries to register events on the input. This is not matching
  // our concept of abstract form field controls. We handle each event manually
  // in "ngDoCheck" based on the form-field control state. The following events
  // need to be handled: focus, blur. We do not handle the "input" event since
  // that one is only needed for the text-field character count, which we do
  // not implement as part of the form-field, but should be implemented manually
  // by consumers using template bindings.
  registerInputInteractionHandler()  {}
  deregisterInputInteractionHandler() {}

  // We do not have a reference to the native input since we work with abstract form field
  // controls. MDC needs a reference to the native input optionally to handle character
  // counting and value updating. These are both things we do not handle from within the
  // form-field, so we can just return null.
  getNativeInput() {
    return null;
  }

  // This method will never be called since we do not have the ability to add event listeners
  // to the native input. This is because the form control is not necessarily an input, and
  // the form field deals with abstract form controls of any type.
  setLineRippleTransformOrigin() {}

  // The foundation tries to register click and keyboard events on the form-field to figure out
  // if the input value changes through user interaction. Based on that, the foundation tries
  // to focus the input. Since we do not handle the input value as part of the form-field, nor
  // it's guaranteed to be an input (see adapter methods above), this is a noop.
  deregisterTextFieldInteractionHandler() {}
  registerTextFieldInteractionHandler() {}

  // The foundation tries to setup a "MutationObserver" in order to watch for attributes
  // like "maxlength" or "pattern" to change. The foundation will update the validity state
  // based on that. We do not need this logic since we handle the validity through the
  // abstract form control instance.
  deregisterValidationAttributeChangeHandler() {}
  registerValidationAttributeChangeHandler() {
    // MDC internally supports "null" if no MutationObserver is registered. The typings of this
    // method limit it to an instance of "MutationObserver" though. We just cast it to "any".
    return null as any;
  }
}

```

Reply:
Ack, that's disapointing. For now, how about just making it a separate function in its own file, then? E.g.
```ts
function getFormFieldAdapter(formField: MatFormField) {
  return {
    // ...
  };
}
```